### PR TITLE
wait_for_service also looks for intra-process srvs

### PR DESCRIPTION
**Pull request by mauropasse**
This is kind of a workaround to have `wait_for_service` work with StubDDS.

There's no such `wait_for_service` API on DDS. It is implemented in `rclcpp` looking for changes in the node graph, using a `wait_set`, which is not supported on StubDDS.

So here, when waiting for a service, we first look into the intra-process services, which doesn't need the wait set. 

If after the timeout it doesn't find it, it'll look with the nodegraph, i.e. will crash since `rmw_wait` is not supported.

The new api `wait_for_intra_process_service_nanoseconds` could become public, so we could directly wait only for intra-process, avoiding the crash if after the timeout the service is not found.

### CODE REVIEW

File: rclcpp/src/rclcpp/client.cpp [hunk 1]
```diff
@@ -119,10 +119,62 @@ ClientBase::service_is_ready() const
...
+
 bool
 ClientBase::wait_for_service_nanoseconds(std::chrono::nanoseconds timeout)
 {
   auto start = std::chrono::steady_clock::now();
+
+  std::lock_guard<std::recursive_mutex> lock(ipc_mutex_);
+  if (use_intra_process_) {
+    // Look first for intra-process services until timeout
+    bool service_available = wait_for_intra_process_service_nanoseconds(timeout);
```
Comment (alsora) [inline rclcpp/src/rclcpp/client.cpp:172]: should this be inside the `do_while` ?
moreover we should release the mutex when we don't need it

**Comment by alsora** [comment]
I'm looking at `rclcpp/src/rclcpp/node_interfaces/node_graph.cpp`.
Do you think that it would be complex to trigger the `graph_cv` from the stub dds?

File: rclcpp/src/rclcpp/client.cpp (hunk 1 above)
Comment (mauropasse) [reply rclcpp/src/rclcpp/client.cpp:172]: You mean the `if (use_intra_process_) {` inside `wait_for_intra_process_service_nanoseconds` ?

Comment (alsora) [reply rclcpp/src/rclcpp/client.cpp:172]: yes, with this code you will always first check for IPC for the whole `timeout` duration and then do the same for non-IPC.
if a service does not exist, you will spend double the time waiting for it.

**Comment by mauropasse** [comment]
It'll probably require a good effort to do anything on the StubDDS and waitsets. This is a quick way to test it. 
If we ever decide we want to support more stuff with wait_sets it could be done.

**Pull request merged**